### PR TITLE
Config, Admin, and UI for Frontend Event Dates Lists

### PR DESCRIPTION
**Pull request by tn3rb**
this PR:

- removes the `/src/` folder to simplify namespaces

- adds route detection logic for loading assets for the EDTR as well as hooking into event dates lists on the frontend

- adds a module for facilitating additional actions and filters

- adds new JsonConfig child class for storing frontend event dates list options

- adds new admin form and handler for managing above config

- adds new css and template files for frontend event dates list

#### new admin section on the Event Espresso - Events - Templates tab

ScreenShot_20200924132411

#### resulting UI templates on frontend:

- default (existing UI)

ScreenShot_20200924132815

- box

ScreenShot_20200924132942

- card

ScreenShot_20200924133144

- circle

ScreenShot_20200924133252

- stripe

ScreenShot_20200924132552

### CODE REVIEW

File: domain/entities/config/RecurringEventsConfig.php [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+
+    /**
+     * @param int $number_of_dates
+     */
+    public function setNumberOfDates($number_of_dates = 6)
+    {
+        $number_of_dates = absint($number_of_dates);
+        $number_of_dates = $number_of_dates > 0 ? $number_of_dates : 1;
+        $this->setProperty('number_of_dates', $number_of_dates);
```
Comment (manzoorwanijk) [inline domain/entities/config/RecurringEventsConfig.php:119]: JS be like, why are you mutating the props? 😄

File: domain/ui/rem-dates-list-datetime.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    <?php } ?>
+    <?php if (! empty($description)) { ?>
+        <div class="ee-rem-date__details--description"><?php echo $description; ?></div>
+    <?php } ?>
+    </div>
+    <div class="ee-rem-date__time">
+        <div class="ee-rem-date__time--start"><?php echo $datetime->start_time($time_format); ?></div>
+        <div class="ee-rem-date__time--sep">-</div>
+        <div class="ee-rem-date__time--end"><?php echo $datetime->end_time($time_format); ?></div>
+    </div>
```
Comment (manzoorwanijk) [inline domain/ui/rem-dates-list-datetime.php:27]: May be add `esc_html()` to output? `description` may be an exception though.

**Review by manzoorwanijk — APPROVED** [review]
Looks great! removing `/src` is a good change 👍

File: domain/ui/rem-dates-list-datetime.php (hunk 2 above)
Comment (tn3rb) [reply domain/ui/rem-dates-list-datetime.php:27]: good thinking, but sanitization is already performed by the PHP models and there is no need to escape those values

File: domain/entities/config/RecurringEventsConfig.php (hunk 1 above)
Comment (tn3rb) [reply domain/entities/config/RecurringEventsConfig.php:119]: hey at least I didn't jam this all into one line of code 😆

**Comment by garthkoyle** [comment]
@Pebblo and @joshfeck This is the sort of thing that I want you to see well in advance of being released. Will you please review these settings and layouts and let us know if you have any feedback?

**Comment by tn3rb** [comment]
You'll need to checkout:

- the `rem-dates-list-and-config` branch for EE core

- the `barista-prod` branch for the REM add-on

that core branch may get merged soon - but I'll update this issue when that happens so you know what branch to use

**Pull request merged**